Add new setting to skip dhc and go straight to Vaati

eternalcode0 opened the pull request:
A new option has been added to the DHC Opens dropdown, "Never, Vaati Warp" which replaces the transition between pedestal and sanctuary, taking you straight to the altar to begin the fight with Vaati. This is a feature that TMCAP (The Minish Cap Archipelago) will be implementing soon as well.

This is just the simplest way I could think to add the Vaati warp into TMCR. TMCAP is separating the dropdown for DHC Opens into a toggle for "Defeat Vaati" and the same dropdown that exists today for "DHC Opens". We do this because we'll eventually have support for all 6 combinations of settings this would make.

So far, I have not done any full test runs to see if everything still works as intended. I've only verified that the transition works properly and that the spoiler log for each dropdown option looks correct.

eternalcode0 committed:
Add new setting to skip dhc and go straight to Vaati

Catobat commented on RandomizerCore/Resources/default.logic:
```diff
@@ -106,7 +106,7 @@
 !flag - Main Settings - Setting - Dungeon Settings - DHC_BARREN - DHC is Barren - If enabled will completely fill the interior of Dark Hyrule Castle with junk aside from dungeon items that have to be placed in there because of the dungeon item settings.\n\nFor technical reasons, dungeon items that are allowed to be placed outside of DHC (and its region) will always be placed outside DHC. Also, se ...
 
-!dropdown - Main Settings - Setting - Requirements - DHC_SETTING - Dark Hyrule Castle Opens - Dark Hyrule Castle is the last dungeon in the game, it has the final boss Vaati. - OPENDHC - Never Open - NODHC - 'Never Open': A barrier is placed at the front of the dungeon, The game ends when you visit the sanctuary after completing all the requirements. - After Requirements - NORMALDHC - 'After Requ ...
+!dropdown - Main Settings - Setting - Requirements - DHC_SETTING - Dark Hyrule Castle Opens - Dark Hyrule Castle is the last dungeon in the game, it has the final boss Vaati. - OPENDHC - Never Open - NODHC - 'Never Open': A barrier is placed at the front of the dungeon, The game ends when you visit the sanctuary after completing all the requirements. - Never, Vaati Warp - FASTVAATI - 'Never, Vaat ...
```
I'd prefer a different name for the final room here since the area is already called "Elemental Sanctuary". Maybe something like "stained glass room"?

Catobat commented on RandomizerCore/Resources/default.logic:
```diff
@@ -8257,9 +8261,13 @@ DHC_3F_SouthEast_Chest		`DHC_INSIDE_SK`;	`DHCCHEST`;	0x88-0x04-0x00;	(|(&`ENTERD
 !ifdef - NODHC
-	BeatVaati;	Helper;	;	`ENTERDHC`
+	!ifdef - FASTVAATI
+		BeatVaati;	Helper;	;	Helpers.CanSplit4, Helpers.HasBow, Items.GustJar, Items.PacciCane, Helpers.DarkRooms
```
This is missing the `ENTERDHC`, part (as seen in the line for the DHC modes). That is responsible for checking reachability of the castle entrance and the pedestal requirements.

Catobat reviewed (changes requested):
Everything else looks good. Thanks for making this!

Catobat commented:
If you go to Vaati and die (or quickwarp), you get sent back to the room with the altar which forces you in the fight again. The only way to escape (aside from defeating him) seems to be to warp home. This is not ideal, though maybe acceptable for now if we force homewarp active when using this option. But this is at the very least worth investigating a bit. Maybe we can have the transition not change the quickwarp point at all, or have it set it to the pedestal. I am not sure how to do either of those, but I could maybe help with looking into it.

Catobat commented:
> If you go to Vaati and die (or quickwarp), you get sent back to the room with the altar which forces you in the fight again. The only way to escape (aside from defeating him) seems to be to warp home. This is not ideal, though maybe acceptable for now if we force homewarp active when using this option. But this is at the very least worth investigating a bit. Maybe we can have the transition not change the quickwarp point at all, or have it set it to the pedestal. I am not sure how to do either of those, but I could maybe help with looking into it.

The problem is that the `locationIndex` of the destination area does not line up with the sanctuary area and so the game updates the respawn location. I think the easiest way to fix this is by setting the `locationIndex` of the sanctuary to the one from DHC. You can do that with `ORG $127649; BYTE 0x1D`. I have not tested the full fight with this, but this at least lets you quickwarp out of the first phase.

eternalcode0 committed:
Fast Vaati fixes

- Change quick warp location for sanctuary
- Fix logic so it correctly requires the ped items
- Change the description for the "Never, Vaati Warp" option

eternalcode0 commented:
Updated the locationIndex when the option is enabled. I tested quick warping out of each of the 3 vaati phases (as well as the 3 rooms in between the 2nd and 3rd phase) and they all correctly send you back into sanctuary without triggering the fight again.

The one minor issue this also introduces is that it makes the game display "Dark Hyrule Castle" upon entering sanctuary instead of "Elemental Sanctuary". I think this is acceptable though. It could maybe be fixed by copying the text for each language's sanctuary region name onto the DHC region name though I'm not sure if that would introduce other issues since all of the translation words are spaced apart by exactly a single 0 byte, the difference in character length between sanctuary and DHC might throw things off.
